src/requests.rs

```diff
@@ -0,0 +1,236 @@
+use std::{thread, time::Duration};
+
+use log::{debug, error, info};
+use reqwest::{blocking::Client, header::CONTENT_TYPE};
+
+use crate::{
+    data::{
+        random_gen_dust_concentration, random_gen_humidity, random_gen_pressure,
+        random_gen_temperature, AirPurity, Reading,
+    },
+    error::RequestSchedulerError,
+    URL,
+};
+
+const DEFAULT_REQUEST_AMOUNT: u32 = 1;
+const DEFAULT_TIME_PER_REQUEST: Duration = Duration::from_secs(10);
+const DEFAULT_NUM_THREADS: u32 = 1;
+const MAX_NUM_THREADS: u32 = 10;
+
+#[derive(Clone, Copy)]
+pub struct RequestSchedulerBuilder {
+    request_amount: Option<u32>,
+    time_per_request: Option<Duration>,
+    total_time: Option<Duration>,
+    num_threads: Option<u32>,
+}
+
+impl RequestSchedulerBuilder {
+    pub fn default() -> Self {
+        RequestSchedulerBuilder {
+            request_amount: None,
+            time_per_request: None,
+            total_time: None,
+            num_threads: None,
+        }
+    }
+
+    pub fn with_some_request_amount(mut self, request_amount: &Option<u32>) -> Self {
+        self.request_amount = *request_amount;
+        self
+    }
+
+    pub fn with_some_time_per_request(mut self, time_per_request: &Option<Duration>) -> Self {
+        self.time_per_request = *time_per_request;
+        self
+    }
+
+    pub fn with_some_total_time(mut self, total_time: &Option<Duration>) -> Self {
+        self.total_time = *total_time;
+        self
+    }
+
+    pub fn with_some_num_threads(mut self, num_threads: &Option<u32>) -> Self {
+        self.num_threads = *num_threads;
+        self
+    }
+
+    pub fn build(self) -> Result<RequestScheduler, RequestSchedulerError> {
+        // Loop indefinitely if no req amt is set. If time per req is also not set then don't loop.
+        let loop_indefinitely = self.request_amount.is_none() && self.time_per_request.is_some();
+
+        let request_amount = self.request_amount.unwrap_or(DEFAULT_REQUEST_AMOUNT);
+
+        let time_per_request = match (&self.time_per_request, &self.total_time) {
+            (None, None) => DEFAULT_TIME_PER_REQUEST,
+            (None, Some(total_time)) => *total_time / request_amount,
+            (Some(time_per_request), None) => *time_per_request,
+            (Some(time_per_request), Some(_)) => *time_per_request,
+        };
+
+        let num_threads = match self.num_threads {
+            Some(num_threads) => num_threads,
+            None => DEFAULT_NUM_THREADS,
+        };
+
+        // Make sure that the number of threads is in [1, `MAX_NUM_THREADS`].
+        match num_threads {
```

Comment:
Is it possible to handle this sort of validation with Clap directly?

Reply:
[It turns out it is yes](https://docs.rs/clap/latest/clap/error/enum.ErrorKind.html#variant.ValueValidation), didn't know :)